src/forms/form-fields/helpers/IconPickerDialog/IconPickerCustomTab.js

```diff
@@ -0,0 +1,210 @@
+import { uploadIcon } from './uploadIcon.js';
+import ErrorMessage from 'd2-ui/lib/messages/ErrorMessage.component';
+import classes from 'classnames';
+
+const styles = {
+    selectFileButton: {
+        backgroundColor: '#ff9800',
+        color: '#fff',
+        textAlign: 'center',
+        position: 'relative',
+        minWidth: 129,
+        height: 36,
+        lineHeight: 2.5,
+        boxShadow: '0 1px 6px rgba(0,0,0,0.12),0 1px 4px rgba(0,0,0,0.12)',
+        cursor: 'pointer',
+        textTransform: 'uppercase',
+    },
+    iconFileInput: {
+        display: 'none',
+    },
+};
+
+export class IconPickerCustomTab extends Component {
+    constructor(props, context) {
+        super(props, context);
+
+        this.t = context.d2.i18n.getTranslation.bind(context.d2.i18n);
+        this.d2 = context.d2;
+        this.state = {
+            iconFile: null,
+            iconFileBase64: null,
+            iconMetadata: {
+                key: '',
+                description: '',
+                keywords: [],
+            },
+            uploading: false,
+            uploadError: null,
+        };
+    }
+
+    handleIconMetadataChange = event => {
+        const value = event.target.value && event.target.value.trim();
+        const name = event.target.name;
+
+        this.setState(prevState => ({
+            iconMetadata: {
+                ...prevState.iconMetadata,
+                [name]: value,
+            },
+        }));
+    };
+
+    handleIconKeywordsChange = event => {
+        const keywords = event.target.value
+            .split(',')
+            .map(kw => kw.trim())
+            .filter(kw => !!kw);
+        this.setState(prevState => ({
+            iconMetadata: {
+                ...prevState.iconMetadata,
+                keywords,
+            },
+        }));
+    };
+
+    handleFileUpload = () => {
+        const iconData = this.state.iconMetadata;
+        this.setState({ uploading: true });
+        uploadIcon(this.state.iconFile, iconData)
+            .then(() => {
+                if (typeof this.props.onIconUpload === 'function') {
+                    this.props.onIconUpload(iconData.key);
+                }
+                this.setState({ uploading: false, uploadError: null });
+            })
+            .catch(e => this.setState({ uploading: false, uploadError: e }));
+    };
+
+    handleFileChange = async event => {
+        const file = event.target.files[0];
+        if (!file) {
+            return;
+        }
+
+        // used to preview image
+        // since this is async, we can't do it during render
+        const iconFileBase64 = await fileToBase64(file);
+
+        this.setState(prevState => ({
+            iconFile: file,
+            iconFileBase64,
+            iconMetadata: {
+                ...prevState.iconMetadata,
+                // remove file extension and replace hyphens with underscores
+                key: file.name.replace(/\..*$/, '').replaceAll('-', '_'),
```

Comment:
maybe remove spaces too since the API doesn't like them, and files often have them

Reply:
Added `\s` and `()` to be replaced.